In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
train_x = pd.read_csv("train_X.csv")
train_y = pd.read_csv("train_y.csv")
test_x = pd.read_csv("test_X.csv")
trainx20= pd.read_csv("trainx20.csv")
trainy20= pd.read_csv("trainy20.csv")
test_x1 = test_x.drop(['Material', 'Id'], axis=1)
train_x1 = train_x.drop(['Material', 'Id'], axis=1)
train_x1

D1  D2  D3       D4        D5  D6  D7  D8  D9      D10  ...     D123  \
0      8  39  31  15.6667  10.22220   8  12  87  75  67.6667  ...  0.00000   
1      9  56  47  17.6250  12.43750   9   9  93  84  77.6250  ...  0.00000   
2      3  46  43  13.6000  12.96000   3   1  87  86  47.6000  ...  0.00000   
3     17  37  20  29.5714   7.34694  17   4  95  91  64.2857  ...  0.00000   
4      8  19  11  12.0000   4.00000   8   3  87  84  62.5000  ...  0.00000   
...   ..  ..  ..      ...       ...  ..  ..  ..  ..      ...  ...      ...   
9635   8  37  29  21.8182  12.56200   8   4  87  83  51.0000  ...  2.11066   
9636   8  31  23  19.0000  11.00000   8  64  87  23  78.0000  ...  0.00000   
9637   8  38  30  12.9333   6.57778   8   8  87  79  75.7333  ...  0.00000   
9638  14  38  24  30.0000  10.66670  38   8  78  70  31.3333  ...  0.00000   
9639  17  37  20  22.3333   7.11111  17   4  94  90  69.3333  ...  0.00031   

          D124      D125  D126  D127  D128  D129     D130        D131  D132  
0     0.000000  0.000000   0.0    12   229   217   78.500   88.666700    12  
1     0.000000  0.000000   0.0    15   229   214   81.125   90.921900    15  
2     0.000000  0.000000   0.0    12   229   217  141.400  103.520000    12  
3     0.000000  0.000000   0.0    64   229   165  134.143   80.163300    64  
4     0.000000  0.000000   0.0    12   229   217  119.500  107.500000    12  
...        ...       ...   ...   ...   ...   ...      ...         ...   ...  
9635  0.383757  0.627966   0.0    12   229   217  130.364  107.603000    12  
9636  0.000000  0.000000   0.0    12   225   213   78.250   73.375000    12  
9637  0.000000  0.000000   0.0     2   229   227   53.800   69.546700    12  
9638  0.000000  0.000000   0.0   225   227     2  225.667    0.888889   225  
9639  0.000034  0.000061   0.0    64   229   165  117.667   71.555600    64  

[9640 rows x 132 columns]

In [ ]:
from sklearn.model_selection import cross_val_score, KFold
from sklearn.metrics import make_scorer, mean_absolute_error, r2_score
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.neural_network import MLPRegressor
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import make_pipeline
from sklearn.svm import SVR
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.preprocessing import StandardScaler
from xgboost import XGBRegressor
'''

scaler = StandardScaler()
X_scaled = scaler.fit_transform(top_100)
'''
# Models to test
models = {
    'XGBoost': XGBRegressor(n_jobs=-1)
}

# Cross-validation settings
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Custom scorer for MAE
mae_scorer = make_scorer(mean_absolute_error, greater_is_better=False)

# Evaluate each model
for name, model in models.items():
    cv_mae = cross_val_score(model, train_x3, train_y2, cv=kf, scoring=mae_scorer, n_jobs=-1)
    cv_r2 = cross_val_score(model, train_x3, train_y2, cv=kf, scoring='r2', n_jobs=-1)
    print(f"{name}: Mean MAE: {-np.mean(cv_mae):.4f}, Mean R2 Score: {np.mean(cv_r2):.4f}")


XGBoost: Mean MAE: 0.5672, Mean R2 Score: 0.7419


## Row-wise Observations

### Found duplicated rows

In [ ]:
import pandas as pd

# Group by all columns
grouped = train_x1.groupby(train_x1.columns.tolist())

# Initialize a dictionary to store indices of each group
group_indices = {}

# Iterate through groups and store indices
group_number = 0
for _, group in grouped:
    if len(group) > 1:  # Only consider groups with more than one row
        group_indices[group_number] = group.index.tolist()
        group_number += 1

# Print or use the group indices
print(group_indices)

# Initialize a dictionary to store Egap values for each group
grouped_egap_values = {}

for group_num, indices in group_indices.items():
    # Extract corresponding Egap values for the group of indices
    grouped_egap_values[group_num] = train_y.loc[indices, 'Egap'].tolist()

# Print or use the grouped Egap values
print(grouped_egap_values)


{0: [3566, 7838], 1: [2848, 3086, 7247, 7397], 2: [270, 2299, 9369], 3: [5570, 7158], 4: [466, 6860], 5: [1876, 2468], 6: [6853, 6915, 7206], 7: [1499, 2603], 8: [5684, 9474], 9: [801, 7432], 10: [162, 474, 2463, 2486, 7825], 11: [1322, 8723], 12: [8996, 9014], 13: [1331, 4358, 4485, 5244, 6385], 14: [322, 5378, 8079], 15: [4653, 6373, 9089], 16: [1762, 5455, 6282, 7932, 9165], 17: [414, 5951], 18: [1248, 8892, 8933], 19: [145, 3393], 20: [8666, 9179], 21: [5057, 7987], 22: [4103, 5996], 23: [1844, 1861], 24: [2281, 8648], 25: [6121, 6289, 7710, 8703], 26: [3759, 6723], 27: [3037, 5595], 28: [1423, 5285], 29: [6564, 9331], 30: [2231, 2305, 3712], 31: [1056, 1809, 5528, 7818], 32: [1005, 1988, 2449, 6897], 33: [3171, 3657, 6847], 34: [46, 5231], 35: [4351, 5635], 36: [3011, 8618], 37: [5347, 6906], 38: [2941, 3841], 39: [3605, 3829], 40: [3015, 9424], 41: [1692, 2056], 42: [2456, 3054], 43: [884, 9361], 44: [3218, 4889, 6052, 6474, 6501, 6603], 45: [1630, 2367], 46: [1175, 6856], 47: [7

### Average their Egap and remove duplicates

In [ ]:
# Initialize a dictionary to store the average Egap for each group
average_egap_per_group = {}

for group_num, egap_values in grouped_egap_values.items():
    # Calculate the average Egap for the group
    average_egap_per_group[group_num] = sum(egap_values) / len(egap_values) if egap_values else 0

# Print or use the average Egap values per group
print(average_egap_per_group)


{0: 6.2445, 1: 4.0020750000000005, 2: 6.7893, 3: 5.0777, 4: 6.00335, 5: 2.36555, 6: 4.672766666666667, 7: 5.53055, 8: 4.8354, 9: 3.25405, 10: 6.411880000000001, 11: 3.5891, 12: 5.05115, 13: 5.70188, 14: 2.7786333333333335, 15: 2.9553999999999996, 16: 4.9846, 17: 4.9155, 18: 3.8263999999999996, 19: 5.359, 20: 2.61825, 21: 3.3309499999999996, 22: 5.59405, 23: 2.7777, 24: 5.2833000000000006, 25: 6.2836, 26: 3.75395, 27: 2.54745, 28: 4.8985, 29: 2.00855, 30: 2.8523, 31: 5.348675, 32: 5.035525, 33: 2.3148666666666666, 34: 1.823, 35: 2.899, 36: 0.49134999999999995, 37: 1.66445, 38: 4.53425, 39: 3.7601500000000003, 40: 4.7052, 41: 2.2022, 42: 2.0069500000000002, 43: 2.7286, 44: 2.977, 45: 2.04405, 46: 3.6089, 47: 2.8554500000000003, 48: 3.2493999999999996, 49: 1.109, 50: 3.37685, 51: 4.90185, 52: 1.85115, 53: 4.384885714285714, 54: 4.3956, 55: 4.73215, 56: 5.35075, 57: 0.15289999999999998, 58: 4.850980000000001, 59: 4.95528, 60: 7.3347999999999995, 61: 3.72675, 62: 3.29665, 63: 2.74845, 64: 2

In [ ]:
train_y1= train_y.copy()
for group_num, indices in group_indices.items():
    # Get the average Egap for this group
    avg_egap = average_egap_per_group[group_num]

    # Update the Egap values for the rows in this group
    train_y1.loc[indices, 'Egap'] = avg_egap


In [ ]:
# Initialize an empty list to store the indices of rows to drop
rows_to_drop = []

# Iterate through the group_indices
for indices in group_indices.values():
    if len(indices) > 1:
        # Add all indices except the first one to the rows_to_drop list
        rows_to_drop.extend(indices[1:])

# Drop these rows from train_x1 and train_y1
train_x2 = train_x1.drop(rows_to_drop)
train_y2 = train_y1.drop(rows_to_drop)
train_y2 = train_y2.drop(['Id'], axis=1)

train_x2 = train_x2.reset_index(drop=True)
train_y2 = train_y2.reset_index(drop=True) #droped duplicates


### Finding highly correlated rows

In [ ]:
import xgboost as xgb

dtrain = xgb.DMatrix(train_x2, label=train_y2)

# Define XGBoost parameters
params = {
    'objective': 'reg:squarederror',  # Use 'binary:logistic' for classification
    'max_depth': 8,
    'min_child_weight': 2.1,
    'eta': 0.3,
    'subsample': 0.95,
    'colsample_bytree': 0.65
}

# Train the model (adjust num_boost_round as needed)
bst = xgb.train(params, dtrain, num_boost_round=10)
importance = bst.get_score(importance_type='gain')
sorted_importance = sorted(importance.items(), key=lambda x: x[1], reverse=True)

feature_importances_df = pd.DataFrame(sorted_importance, columns=['Feature', 'Importance'])

drop = feature_importances_df[feature_importances_df['Importance']<4]['Feature'].tolist()
train_x3 = train_x2.drop(columns=drop)
train_x3 # dropped least important features
test_x2 = test_x1.drop(columns=drop)
test_x2

D1  D2  D3       D4       D5  D6  D7  D8  D9      D10  ...  D123  D124  \
0      6  79  73  23.6667  22.8889   6   4  82  78  64.6667  ...   0.0   0.0   
1      6  11   5   7.0000   1.0000   6   2  82  80  69.5000  ...   0.0   0.0   
2      6   6   0   6.0000   0.0000   6  77  77   0  77.0000  ...   0.0   0.0   
3      1  16  15  10.0000   4.4000   9  83  93  10  89.8000  ...   0.0   0.0   
4      6  56  50  22.6667  22.2222   6   9  77  68  54.3333  ...   0.0   0.0   
...   ..  ..  ..      ...      ...  ..  ..  ..  ..      ...  ...   ...   ...   
4127   9  83  74  26.0000  22.6667   9   4  93  89  77.0000  ...   0.0   0.0   
4128  14  14   0  14.0000   0.0000  14  78  78   0  78.0000  ...   0.0   0.0   
4129   8  57  49  25.8333  12.8056  16  13  88  75  71.5000  ...   0.0   0.0   
4130  15  79  64  27.0000  17.3333  17  66  94  28  87.5000  ...   0.0   0.0   
4131   3  23  20   8.0000   3.7500   8   1  87  86  49.6250  ...   0.0   0.0   

      D125  D126  D127  D128  D129      D130       D131  D132  
0      0.0   0.0   194   229    35  205.0000   14.66670   194  
1      0.0   0.0   194   229    35  198.3750    7.65625   194  
2      0.0   0.0   194   194     0  194.0000    0.00000   194  
3      0.0   0.0     2   194   192   59.2000   58.24000    15  
4      0.0   0.0   194   229    35  205.6670   15.55560   194  
...    ...   ...   ...   ...   ...       ...        ...   ...  
4127   0.0   0.0    12   229   217   50.1667   59.61110    15  
4128   0.0   0.0   227   227     0  227.0000    0.00000   227  
4129   0.0   0.0    12   194   182   95.1667   49.41670    70  
4130   0.0   0.0     2   225   223   80.5000   48.16670    64  
4131   0.0   0.0    12   229   217  120.5000  108.50000    12  

[4132 rows x 125 columns]

In [ ]:
Ttrain_x3 = train_x3.T
# Assuming the rest of your code is the same
correlation_matrix = Ttrain_x3.corr().abs()
high_corr = (correlation_matrix.where(np.triu(np.ones(correlation_matrix.shape), k=1).astype(bool))
             .stack()
             .sort_values(ascending=False))
highly_correlated_pairs = high_corr[high_corr > 0.99997]
print(highly_correlated_pairs)


correlated_pairs_list = list(highly_correlated_pairs.index)
#print(correlated_pairs_list)

merged_groups = []

for pair in correlated_pairs_list:
    merged = False
    for group in merged_groups:
        # Check if the pair shares any element with the current group
        if set(pair).intersection(set(group)):
            # Merge the pair with this group
            group.update(pair)
            merged = True
            break
    if not merged:
        # If the pair wasn't merged with any existing group, add as a new group
        merged_groups.append(set(pair))

# Convert sets back to lists (optional)
merged_groups = [list(group) for group in merged_groups]

grouped_egap_values = {}

for group_num, indices in enumerate(merged_groups):
    # Extract corresponding Egap values for the group of indices
    grouped_egap_values[group_num] = train_y2.loc[indices, 'Egap'].tolist()

# Print or use the grouped Egap values
print(grouped_egap_values)
# Initialize a dictionary to store the average Egap for each group
average_egap_per_group = {}

for group_num, egap_values in grouped_egap_values.items():
    # Calculate the average Egap for the group
    average_egap_per_group[group_num] = sum(egap_values) / len(egap_values) if egap_values else 0

# Print or use the average Egap values per group
print(average_egap_per_group)


249   1477    1.00000
5761  7492    1.00000
652   1477    1.00000
249   652     1.00000
77    4203    1.00000
               ...   
889   3340    0.99997
2586  4041    0.99997
3053  5436    0.99997
4862  5965    0.99997
31    6936    0.99997
Length: 622, dtype: float64
{0: [1.9154, 2.545, 2.1033], 1: [1.4513, 1.5406], 2: [1.0767, 2.9347999999999996], 3: [3.8518, 4.1206], 4: [3.4602, 3.5643, 3.5028], 5: [3.7273, 3.40885], 6: [5.1227, 4.6858, 2.8143], 7: [0.1664, 0.5992, 1.0541, 0.5727, 0.2815, 0.2853, 0.5381], 8: [3.0284, 3.8462, 2.8705, 3.5527], 9: [0.5567, 1.51125], 10: [1.0897, 1.0981, 0.9783], 11: [0.0912, 1.1634], 12: [1.222, 0.0924], 13: [0.7891, 1.0], 14: [2.5386, 2.2588, 2.2771, 2.8874], 15: [4.0141, 2.8977], 16: [6.5369, 2.5937], 17: [3.3835, 3.5096], 18: [2.8853999999999997, 0.42560000000000003], 19: [4.0922, 3.9238, 3.5351, 3.3957], 20: [2.2925, 2.1244, 2.5988], 21: [3.214, 3.8855], 22: [3.9703, 2.0609], 23: [3.6888, 1.8551, 3.555, 3.4796], 24: [1.1699, 4.0451], 25: [0.1555, 

In [ ]:
train_y3 = train_y2.copy()

# Update the Egap values for each group
for group_num, indices in enumerate(merged_groups):
    avg_egap = average_egap_per_group[group_num]
    train_y3.loc[indices, 'Egap'] = avg_egap

# Initialize an empty list to store the indices of rows to drop
rows_to_drop = []

# Corrected loop to build rows_to_drop
for _, indices in enumerate(merged_groups):
    if len(indices) > 1:
        # Add all indices except the first one to the rows_to_drop list
        rows_to_drop.extend(indices[1:])

# Drop these rows from train_x3 and train_y3
train_x4 = train_x3.drop(rows_to_drop)
train_y4 = train_y3.drop(rows_to_drop)

# Reset the indices
train_x4.reset_index(drop=True, inplace=True)
train_y4.reset_index(drop=True, inplace=True)


## Columnwise Observations

In [ ]:
unique_values_count = train_x4.nunique()
unique_values_count

# Filter columns where the number of unique values is exactly 1
columns_with_single_unique_value = unique_values_count[unique_values_count == 1].index

print("Columns with exactly one unique value and the unique value:")
for column in columns_with_single_unique_value:
    # Find the unique value in the column
    unique_value = train_x[column].unique()[0]
    print(f"{column}: {unique_value}")

train_x4 = train_x4.drop(columns_with_single_unique_value, axis=1)
test_x2= test_x2.drop(columns_with_single_unique_value, axis=1)
train_x4

Columns with exactly one unique value and the unique value:
D97: 0
D121: 0


D1  D2  D3       D4        D5  D6  D7  D8  D9      D10  ...     D123  \
0      8  39  31  15.6667  10.22220   8  12  87  75  67.6667  ...  0.00000   
1      9  56  47  17.6250  12.43750   9   9  93  84  77.6250  ...  0.00000   
2      3  46  43  13.6000  12.96000   3   1  87  86  47.6000  ...  0.00000   
3     17  37  20  29.5714   7.34694  17   4  95  91  64.2857  ...  0.00000   
4      8  19  11  12.0000   4.00000   8   3  87  84  62.5000  ...  0.00000   
...   ..  ..  ..      ...       ...  ..  ..  ..  ..      ...  ...      ...   
7898   5  38  33  13.8333   9.75000   8   8  87  79  75.7500  ...  0.00000   
7899   8  37  29  21.8182  12.56200   8   4  87  83  51.0000  ...  2.11066   
7900   8  38  30  12.9333   6.57778   8   8  87  79  75.7333  ...  0.00000   
7901  14  38  24  30.0000  10.66670  38   8  78  70  31.3333  ...  0.00000   
7902  17  37  20  22.3333   7.11111  17   4  94  90  69.3333  ...  0.00031   

          D124      D125  D126  D127  D128  D129      D130        D131  D132  
0     0.000000  0.000000   0.0    12   229   217   78.5000   88.666700    12  
1     0.000000  0.000000   0.0    15   229   214   81.1250   90.921900    15  
2     0.000000  0.000000   0.0    12   229   217  141.4000  103.520000    12  
3     0.000000  0.000000   0.0    64   229   165  134.1430   80.163300    64  
4     0.000000  0.000000   0.0    12   229   217  119.5000  107.500000    12  
...        ...       ...   ...   ...   ...   ...       ...         ...   ...  
7898  0.000000  0.000000   0.0    12   225   213   64.0833   60.791700    12  
7899  0.383757  0.627966   0.0    12   229   217  130.3640  107.603000    12  
7900  0.000000  0.000000   0.0     2   229   227   53.8000   69.546700    12  
7901  0.000000  0.000000   0.0   225   227     2  225.6670    0.888889   225  
7902  0.000034  0.000061   0.0    64   229   165  117.6670   71.555600    64  

[7903 rows x 123 columns]

### Dropping High correlated columns

In [ ]:
import pandas as pd

# Set the option to display more rows
pd.set_option('display.max_rows', None)

# Assuming the rest of your code is the same
correlation_matrix = train_x4.corr().abs()
high_corr = (correlation_matrix.where(np.triu(np.ones(correlation_matrix.shape), k=1).astype(bool))
             .stack()
             .sort_values(ascending=False))
highly_correlated_pairs = high_corr[high_corr > 0.8]
#print(highly_correlated_pairs)

pd.reset_option('display.max_rows')
columns_to_drop = set()

for (col1, col2), corr_value in highly_correlated_pairs.items():
    if corr_value > 0.97:
        # Add one of the pair to the set of columns to drop
        columns_to_drop.add(col1)  # You can also choose col2

# Convert the set to a list, if needed
columns_to_drop = list(columns_to_drop)
train_x4_dropped = train_x4.drop(columns=columns_to_drop)
test_x1_dropped = test_x2.drop(columns=columns_to_drop)
train_x4_dropped

D8  D9      D10      D11  D12      D13       D14       D15      D16  \
0     87  75  67.6667  25.7778   87  15.9994   88.9059   72.9065  33.9742   
1     93  84  77.6250  21.1406   93  18.9984  137.3270  118.3290  39.8569   
2     87  86  47.6000  37.2800    1   6.9410  106.4200   99.4790  30.4602   
3     95  91  64.2857  34.5306    4  35.4530   85.4678   50.0148  66.4565   
4     87  84  62.5000  29.7500   87  15.9994   39.0983   23.0989  24.5197   
...   ..  ..      ...      ...  ...      ...       ...       ...      ...   
7898  87  79  75.7500  13.1250   87  10.8110   87.6200   76.8090  30.0570   
7899  87  83  51.0000  34.1818   87  15.9994   85.4678   69.4684  48.5053   
7900  87  79  75.7333  16.9600   87  15.9994   87.6200   71.6206  27.4296   
7901  78  70  31.3333  31.1111    8  28.0855   87.6200   59.5345  67.7752   
7902  94  90  69.3333  32.8889   94  35.4530   85.4678   50.0148  48.7324   

           D17  ...   D120     D123      D125  D126  D127  D128  D129  \
0     23.96630  ...  0.000  0.00000  0.000000   0.0    12   229   217   
1     30.28980  ...  1.970  0.00000  0.000000   0.0    15   229   214   
2     30.38390  ...  0.000  0.00000  0.000000   0.0    12   229   217   
3     18.54790  ...  0.000  0.00000  0.000000   0.0    64   229   165   
4      8.52026  ...  0.000  0.00000  0.000000   0.0    12   229   217   
...        ...  ...    ...      ...       ...   ...   ...   ...   ...   
7898  22.81580  ...  0.000  0.00000  0.000000   0.0    12   225   213   
7899  29.55080  ...  0.000  2.11066  0.627966   0.0    12   229   217   
7900  15.24030  ...  0.000  0.00000  0.000000   0.0     2   229   227   
7901  26.45980  ...  0.000  0.00000  0.000000   0.0   225   227     2   
7902  17.70590  ...  2.493  0.00031  0.000061   0.0    64   229   165   

          D130        D131  D132  
0      78.5000   88.666700    12  
1      81.1250   90.921900    15  
2     141.4000  103.520000    12  
3     134.1430   80.163300    64  
4     119.5000  107.500000    12  
...        ...         ...   ...  
7898   64.0833   60.791700    12  
7899  130.3640  107.603000    12  
7900   53.8000   69.546700    12  
7901  225.6670    0.888889   225  
7902  117.6670   71.555600    64  

[7903 rows x 102 columns]

### Drop columns with mostly zero

In [ ]:
import pandas as pd

# Set the option to display more rows, if not already set
pd.set_option('display.max_rows', None)

# Assuming your data is in a DataFrame `top_100_dropped`
# Calculate the percentage of zeros in each column
percent_zeros = train_x4_dropped.apply(lambda col: (col == 0).mean() * 100)

# Display the results
#print(percent_zeros)

# Reset the display option if needed
pd.reset_option('display.max_rows')
columns_to_drop = percent_zeros[percent_zeros > 88].index.tolist()
top_100_filtered = train_x4_dropped.drop(columns=columns_to_drop)
test_x1_filtered = test_x1_dropped.drop(columns=columns_to_drop)
test_x1_filtered

D8  D9      D10       D11  D12       D13       D14       D15      D16  \
0     82  78  64.6667  20.22220   77  12.01070  196.9670  184.9560  55.7449   
1     82  80  69.5000  16.87500   77  12.01070   22.9898   10.9791  14.1316   
2     77   0  77.0000   0.00000   77  12.01070   12.0107    0.0000  12.0107   
3     93  10  89.8000   3.44000   93   1.00794   32.0650   31.0571  20.4087   
4     77  68  54.3333  30.22220   77  12.01070  137.3270  125.3160  53.7828   
...   ..  ..      ...       ...  ...       ...       ...       ...      ...   
4127  93  89  77.0000  24.33330   93  18.99840  208.9800  189.9820  61.7403   
4128  78   0  78.0000   0.00000   78  28.08550   28.0855    0.0000  28.0855   
4129  88  75  71.5000  19.91670   88  15.99940  138.9050  122.9060  57.5855   
4130  94  28  87.5000   8.66667   94  30.97380  196.9670  165.9930  61.6254   
4131  87  86  49.6250  37.37500   87   6.94100   50.9415   44.0005  16.9703   

           D17  ...      D119   D120  D123  D125  D127  D128  D129      D130  \
0     56.98150  ...  2.429560  4.496   0.0   0.0   194   229    35  205.0000   
1      2.21455  ...  1.331340  4.496   0.0   0.0   194   229    35  198.3750   
2      0.00000  ...  0.000000  4.496   0.0   0.0   194   194     0  194.0000   
3      8.88854  ...  1.891600  1.970   0.0   0.0     2   194   192   59.2000   
4     55.69610  ...  1.998220  4.496   0.0   0.0   194   229    35  205.6670   
...        ...  ...       ...    ...   ...   ...   ...   ...   ...       ...   
4127  56.98920  ...  0.875556  1.970   0.0   0.0    12   229   217   50.1667   
4128   0.00000  ...  0.000000  0.773   0.0   0.0   227   227     0  227.0000   
4129  32.45150  ...  1.101000  2.202   0.0   0.0    12   194   182   95.1667   
4130  45.11370  ...  0.746889  2.493   0.0   0.0     2   225   223   80.5000   
4131   8.49281  ...  0.000000  0.000   0.0   0.0    12   229   217  120.5000   

           D131  D132  
0      14.66670   194  
1       7.65625   194  
2       0.00000   194  
3      58.24000    15  
4      15.55560   194  
...         ...   ...  
4127   59.61110    15  
4128    0.00000   227  
4129   49.41670    70  
4130   48.16670    64  
4131  108.50000    12  

[4132 rows x 88 columns]

## All Model Testing

In [ ]:
from sklearn.model_selection import cross_val_score, KFold
from sklearn.metrics import make_scorer, mean_absolute_error, r2_score
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.neural_network import MLPRegressor
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import make_pipeline
from sklearn.svm import SVR
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from xgboost import XGBRegressor
from lightgbm import LGBMRegressor
from sklearn.neighbors import KNeighborsRegressor


models = {
    'LinearRegression': LinearRegression(),
    'Ridge': Ridge(),
    'Lasso': Lasso(),
    'ElasticNet': ElasticNet(),
    'PolynomialRegression': make_pipeline(PolynomialFeatures(degree=2), LinearRegression()),
    'MLPRegressor': MLPRegressor(max_iter=1000),
    'SVR': SVR(),
    'XGBoost': XGBRegressor(n_jobs=-1),
    'RandomForest': RandomForestRegressor(n_jobs=-1),
    'GradientBoosting': GradientBoostingRegressor(),
    'LightGBM': LGBMRegressor(n_jobs=-1),
    'DecisionTree': DecisionTreeRegressor(),
    'KNN': KNeighborsRegressor()

}

# Cross-validation settings
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Custom scorer for MAE
mae_scorer = make_scorer(mean_absolute_error, greater_is_better=False)

# Evaluate each model
for name, model in models.items():
    cv_mae = cross_val_score(model, train_x2, train_y2, cv=kf, scoring=mae_scorer, n_jobs=-1)
    cv_r2 = cross_val_score(model, train_x2, train_y2, cv=kf, scoring='r2', n_jobs=-1)
    print(f"{name}: Mean MAE: {-np.mean(cv_mae):.4f}, Mean R2 Score: {np.mean(cv_r2):.4f}")


In [ ]:
from sklearn.model_selection import cross_val_score, KFold
from sklearn.metrics import make_scorer, mean_absolute_error, r2_score
from sklearn.pipeline import make_pipeline
from xgboost import XGBRegressor

models = {

    'XGBoost': XGBRegressor(n_jobs=-1)
}

# Cross-validation settings
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Custom scorer for MAE
mae_scorer = make_scorer(mean_absolute_error, greater_is_better=False)

# Evaluate each model
for name, model in models.items():
    cv_mae = cross_val_score(model, top_100_filtered, train_y4, cv=kf, scoring=mae_scorer, n_jobs=-1)
    cv_r2 = cross_val_score(model, top_100_filtered, train_y4, cv=kf, scoring='r2', n_jobs=-1)
    print(f"{name}: Mean MAE: {-np.mean(cv_mae):.4f}, Mean R2 Score: {np.mean(cv_r2):.4f}")


XGBoost: Mean MAE: 0.5722, Mean R2 Score: 0.7434


XGBoost: Mean MAE: 0.5700, Mean R2 Score: 0.7396

train_x4 XGBoost: Mean MAE: 0.5694, Mean R2 Score: 0.7445

train_x4_dropped XGBoost: Mean MAE: 0.5719, Mean R2 Score: 0.7409

top_100_filtered XGBoost: Mean MAE: 0.5722, Mean R2 Score: 0.7434



In [ ]:
train_x4.to_csv('train_x4.csv', index=False)
train_y4.to_csv('train_y4.csv', index=False)
test_x2.to_csv('test_x2.csv', index=False)


## Apparently, XGBoost performs the best. I will then focus on hypertuning XGBoost model.

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import cross_val_score, KFold
from sklearn.metrics import make_scorer, mean_absolute_error, r2_score
from xgboost import XGBRegressor
#from sklearn.preprocessing import StandardScaler

# Define the optimized parameters for XGBoost
xgb_params = {

    'subsample':  0.95,  #0.8
    'n_estimators': int(4500),  # Convert to integer
    'min_child_weight': 2.1,
    'max_depth': int(7),  # Convert to integer
    'learning_rate': 0.02,
    'gamma': 0, #0
    'colsample_bytree': 0.65, #0.8
    'nthread': 4,
    'scale_pos_weight' : 1,
    'reg_alpha': 1e-05,
    'objective':'reg:squarederror'
}


# Models to test
models = {
    'XGBoost': XGBRegressor(**xgb_params)
}
'''scaler = StandardScaler()
X_scaled = scaler.fit_transform(train_x4)'''

# Cross-validation settings
kf = KFold(n_splits=7, shuffle=True, random_state=77)

# Custom scorer for MAE
mae_scorer = make_scorer(mean_absolute_error, greater_is_better=False)

# Evaluate each model using cross-validation
for name, model in models.items():
    cv_mae = cross_val_score(model, train_x4, train_y4, cv=kf, scoring=mae_scorer, n_jobs=-1)
    cv_r2 = cross_val_score(model, train_x4, train_y4, cv=kf, scoring='r2', n_jobs=-1)
    print(f"{name}: Mean MAE: {-np.mean(cv_mae):.4f}, Mean R2 Score: {np.mean(cv_r2):.4f}")

# Train the model on the entire dataset
model = XGBRegressor(**xgb_params)
model.fit(train_x4, train_y4)

# Assuming test_x is your test dataset
# Make sure test_x is prepared in the same way as your training data
# Predict on the test set
predicted_egap = model.predict(test_x2)

# Prepare submission DataFrame
# Replace 'Id' with the actual ID column of your test dataset
submission_df = pd.DataFrame()
submission_df['Id'] = test_x['Id']
submission_df['Egap'] = predicted_egap

# Save predictions to CSV
filename = f"XGBoost_predictions_{-np.mean(cv_mae):.4f}.csv"
submission_df.to_csv(filename, index=False)


/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py:752: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  warnings.warn(


XGBoost: Mean MAE: 0.5095, Mean R2 Score: 0.7852


In [ ]:
xgb_params = {

    'subsample':  0.95,  #0.8
    'n_estimators': int(4500),  # Convert to integer
    'min_child_weight': 2.1,
    'max_depth': int(7),  # Convert to integer
    'learning_rate': 0.02,
    'gamma': 0, #0
    'colsample_bytree': 0.65, #0.8
    'nthread': 4,
    'scale_pos_weight' : 1,
    'reg_alpha': 1e-05,
    'objective':'reg:squarederror'
}XGBoost: Mean MAE: 0.5095, Mean R2 Score: 0.7852
